imports/lib/schemas/typedSchemas.ts

```diff
@@ -0,0 +1,256 @@
+      return buildLiteralUnionField(fieldName, nonOptional, overrides, optional);
+    }
+    return [[
+      fieldName,
+      (<any>SimpleSchema).oneOf(nonOptional.map((subType) => {
+        const fields = buildField(fieldName, subType, overrides);
+        if (fields.length > 1) {
+          throw new Error('array types as a member of a union are not allowed');
+        }
+        return fields[0][1];
+      })),
+    ]];
+  } else if (fieldType instanceof t.StringType) {
+    return [[fieldName, { ...overrides, type: String, optional }]];
+  } else if (
+    // I'm pretty sure this instanceof check shouldn't be necessary, since
+    // RefinementType extends Type, but the compiler seems unhappy with the
+    // comparison otherwise.
+    fieldType instanceof t.RefinementType &&
+    fieldType === t.Integer
+  ) {
+    return [[fieldName, { ...overrides, type: SimpleSchema.Integer, optional }]];
+  } else if (fieldType instanceof t.NumberType) {
+    return [[fieldName, { ...overrides, type: Number, optional }]];
+  } else if (fieldType instanceof t.BooleanType) {
+    return [[fieldName, { ...overrides, type: Boolean, optional }]];
+  } else if (fieldType instanceof DateType) {
+    return [[fieldName, { ...overrides, type: Date, optional }]];
+  } else if (fieldType instanceof t.ArrayType) {
+    const schemaOverrides = _.omit(overrides, 'array');
+    const arrayOverrides = overrides && (<ArrayOverrides<any[]>>overrides).array;
+    return [
+      [fieldName, { ...schemaOverrides, type: Array, optional }],
+      ...buildField(`${fieldName}.$`, fieldType.type, arrayOverrides),
+    ];
+  } else if (fieldType instanceof t.InterfaceType) {
+    const schemaOverrides = _.omit(overrides, 'nested');
+    const nestedOverrides = overrides && (<ObjectOverrides<Record<string, any>>>overrides).nested;
+    return [[fieldName, {
+      ...schemaOverrides,
+      // eslint-disable-next-line no-use-before-define
+      type: buildSchema(fieldType, nestedOverrides || {}),
+      optional,
+    }]];
+  } else if (fieldType instanceof t.ObjectType) {
+    return [[fieldName, {
+      ...overrides,
+      type: Object,
+      blackbox: true,
+      optional,
+    }]];
+  }
+
+  throw new Meteor.Error(`Unknown type ${fieldType.name}`);
+};
+
+export const buildSchema = function <
+  T extends Record<string, any>,
+  P extends Record<keyof T, t.Mixed>
+> (
+  schemaType: t.InterfaceType<P, T>,
+  overrides: Overrides<T>
+): SimpleSchema {
+  const schema: Record<string, FieldDefinition<any>> = {};
+  Object.keys(schemaType.props).forEach((k) => {
+    const fields = buildField(k, schemaType.props[k], overrides[k]);
+    fields.forEach(([name, definition]) => {
+      schema[name] = definition;
+    });
+  });
+
+  return new SimpleSchema(schema);
+};
+
+export const inheritSchema = function <
+  PT extends Record<string, any>,
+  PP extends Record<keyof PT, t.Mixed>,
+  CT extends Record<string, any>,
+  CP extends Record<keyof CT, t.Mixed>
+> (
```

Comment:
Consider writing some comments explaining what properties drive us toward this design, for posterity?